packages/codemods/src/helpers/either.ts

```diff
@@ -0,0 +1,89 @@
+import { Chainable, Flattened } from './chainable';
+
+export interface Right<R, L> extends EitherInternal<R, L> {
+  left: false;
+  right: true;
+  value: R;
+}
+export interface Left<R, L> extends EitherInternal<R, L> {
+  left: true;
+  right: false;
+  value: L;
+}
+
+class EitherInternal<R, L> implements Chainable<R> {
+  public left: boolean;
+  public right: boolean;
+  public value: R | L;
+
+  public constructor(options: { left: boolean; right: boolean; value: R | L }) {
+    this.left = options.left;
+    this.right = options.right;
+    this.value = options.value;
+  }
+
+  public fmap<T>(this: Either<R, L>, fn: (v: R) => T): Either<T, L> {
+    if (this.right) {
+      return Right(fn(this.value));
+    }
+
+    return Left(this.value);
+  }
+
+  public chain<T>(this: Either<R, L>, fn: (v: R) => Either<T, L>): Either<T, L> {
+    if (this.right) {
+      return fn(this.value);
+    }
+    return Left(this.value);
+  }
+
+  public flatten(): Flattened<R, Either<R, L>> {
+    if (this.value && this.value instanceof EitherInternal) {
+      return this.value as Flattened<R, Either<R, L>>;
+    }
+    return (this as unknown) as Flattened<R, Either<R, L>>;
+  }
+
+  public then<F>(this: Either<R, L>, fnR: (v: R) => F | Either<F, L>, fnL?: (v: L) => F | Either<F, L>): Either<F, L> {
```

Comment:
i don't quite understand this `then` method. why support 2 functions here?

Should it be something like
```
if (this.left) {
   return Left(...)
} 

return Right(..)
```

Reply:
This can't support 2 single functions due to the expect input of those lambdas. Having a single function means that the type required to assess is `R | L` which then means that you must know something about the type of the Either which negates a lot of its usefulness. I could potentially rename this whole thing as result. Another way to look at the typings could be `Return Value | Error Value` and you might want to preform several operations on the return value which could return a new Error Value at any time. But the error cannot then return a valid return value.

Reply:
I don't quite follow. my understanding for `then` is to proceed another operation on the same direction `this` currently is on, something like: `public then<F>(this: Either<R, L>, fn: (v: R | L) => F): Either<F, L> | Either<R, L>`

This is why having JSDoc for these methods can really help.
Also we don't have test case that covers all the `then` code paths, therefore I am confused what it's designed to do :)